vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_flash.c

```diff
@@ -0,0 +1,490 @@
+    #include "nrf_fstorage_sd.h"
+#else
+    #include "nrf_drv_clock.h"
+    #include "nrf_fstorage_nvmc.h"
+#endif
+#include "nrf52840_peripherals.h"
+
+/* Main includes. */
+#include "iot_config.h"
+#include "iot_flash.h"
+#include "iot_flash_config.h"
+
+#ifdef IOT_LOG_LEVEL_GLOBAL
+    #define LIBRARY_LOG_LEVEL    IOT_LOG_LEVEL_GLOBAL
+#else
+    #define LIBRARY_LOG_LEVEL    IOT_LOG_LEVEL_IOT_FLASH
+#endif
+
+#define LIBRARY_LOG_NAME         ( "IOT_FLASH" )
+#include "iot_logging_setup.h"
+
+
+/* Flash constants */
+#define FLASH_PAGE_SIZE         0x1000     /* 4k bytes */
+#define FLASH_SECTOR_SIZE       0x1000     /* 4k bytes */
+#define FLASH_BLOCK_SIZE        0x1000     /* 4k bytes */
+#define FLASH_SECTOR_MASK_4K    0x0FFF     /* 4k mask */
+
+#define MAX_WR_SIZE             NRF_FSTORAGE_SD_MAX_WRITE_SIZE
+#define FLASH_PADDING_BYTE      0xFF       /* NAND-flash default logic value */
+
+static void prvFlashEventHandler( nrf_fstorage_evt_t * p_evt );
+
+NRF_FSTORAGE_DEF( nrf_fstorage_t iot_fstorage ) =
+{
+    /* Set a handler for fstorage events. */
+    .evt_handler = prvFlashEventHandler,
+};
+
+typedef enum
+{
+    IOT_FLASH_CLOSED = 0u,
+    IOT_FLASH_OPENED = 1u
+} IotFlashState_t;
+
+/* @brief flash context */
+typedef struct IotFlashDescriptor
+{
+    IotFlashInfo_t xFlashInfo;    /* flash info structure */
+    IotFlashCallback_t xCallback; /* callback */
+    void * pvUserContext;         /* user context to provide in callback */
+    uint8_t ucState;
+} IotFlashDescriptor_t;
+
+#define DEFAULT_FLASH_DESC                          \
+    ( IotFlashDescriptor_t )                        \
+    {                                               \
+        .xFlashInfo =                               \
+        {                                           \
+            .ulFlashSize       = 0,                 \
+            .ulBlockSize       = FLASH_BLOCK_SIZE,  \
+            .ulSectorSize      = FLASH_SECTOR_SIZE, \
+            .ulPageSize        = FLASH_PAGE_SIZE,   \
+            .ulLockSupportSize = 0,                 \
+            .ucAsyncSupported  = 0,                 \
+        },                                          \
+        .xCallback = NULL,                          \
+        .pvUserContext = NULL,                      \
+        .ucState = IOT_FLASH_CLOSED                 \
+    }
+
+static IotFlashDescriptor_t xFlashDesc = DEFAULT_FLASH_DESC;
+
+/*---------------------------------------------------------------------------------*
+*                                Private Helpers                                  *
+*---------------------------------------------------------------------------------*/
+static inline bool prvIsAligned4K( uint32_t ulAddr )
+{
+    return ( ulAddr & FLASH_SECTOR_MASK_4K ) == 0;
+}
+
+static inline void prvWaitUntilFlashReady( nrf_fstorage_t const * p_fstorage )
+{
+    while( nrf_fstorage_is_busy( p_fstorage ) )
+    {
+        vTaskDelay( pdMS_TO_TICKS( 1 ) );
```

Comment:
Haven't read the whole PR, but maybe should return IOT_FLASH_DEVICE_BUSY to the application, and defer waiting to the next level?

Reply:
I agree.

Reply:
By 'next level' do you mean above `iot_flash.h`? If so, and the concern is to not delay/reschedule the current task, I see where you're coming from.  However the NVMC does not have an interrupt based interface so if `vTaskDelay` can't be used then we're limited to busy-waits or delay via hardware timer -- all the while still having to block parent API call as its sync. `prvWaitUntilFlashReady` is only called in the `*_sync` functions which, by API definition, can not return until the flash operation is complete.

Or by next level, do you mean having `prvWaitUntilFlashReady` return `IOT_FLASH_DEVICE_BUSY`?

Reply:
Yes, if the API returns busy to the user, then we defer the decision to them. This allows for more flexibility like yielding the thread or choosing how long to delay. In the case of a library sitting on this one, it can bubble up the BUSY, and make for a more flexible library

Reply:
I think that would work well for `async`, but per the sync API:

[`iot_flash.h#317`](https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/common_io/include/iot_flash.h#L317):
```
 *          This is a blocking operation and waits until the number of bytes are written before returning.
```
[`iot_flash.h#332`](https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/common_io/include/iot_flash.h#L332):
```
IOT_FLASH_DEVICE_BUSY if another asynchronous operation is currently being executed.
```
however async is currently not supported.